docs/graphql.md

```diff
@@ -0,0 +1,102 @@
+# GraphQL transactions and spans
+
+**NB:** This document is a work in progress.
+
+Our current approach of splitting and grouping transaction by HTTP method and path fits perfectly for REST-style APIs and the like.
+
+With a GraphQL approach, the client always requests the same endpoint. This means wildly different queries and mutations all will end up in the same transaction group.
+
+To better support GraphQL servers we need to find another way to tell the queries apart.
+
+Example GraphQL query:
+
+```graphql
+{
+  user {
+    id
+    name
+    comments {
+      body
+    }
+  }
+}
+```
+
+**Note:** The Node.js agent already supports GraphQL. This spec is written with that in mind but not necessarily with its implementation as a target result.
+
+## Prefix
+
+To distinguish GraphQL spans from others we prefix them with `GraphQL:`. A span with just the name `User` will be hard to recognise from just that name, whereas `GraphQL:User` is easy to recognise.
+
+## Operation Name
+
+It is common (and [recommended](https://graphql.org/learn/queries/#operation-name)) to provide an _Operation Name_ for queries. Here for example `UserWithName`:
+
+```graphql
+query UserWithComments {
+  user {
+    id
+    name
+    comments {
+      body
+    }
+  }
+}
+```
+
+The point of these are to provide an easy way for the developers, when things go wrong, to pinpoint where exactly they did so.
+
+This name is available on the server too and serves as a great distinguishing key.
+
+Span name examples:
+- `GraphQL:UserWithComments`
+- `GraphQL:UpdateUser`
+
+## Multiple endpoints
+
+An app may serve multiple GraphQL endpoints. To tell them apart we can include the path in the transaction name.
+
+- `GraphQL:UserWithComments (/api/graphql)`
+
+I'm not sure this is a common thing to do, so perhaps this could be an opt-in option, like `graphql_postfix_path: true`.
+
+## Anonymous queries
+
+An Operation Name isn't required. When one isn't provided it's hard for us to tell apart the queries.
+
+Some clients generate `id`s from hashing the contents of the query (see [apollo-tooling](https://github.com/apollographql/apollo-tooling/blob/1dfd737eaf85b89b2cfb13913342e091e3c03d18/packages/apollo-codegen-core/src/compiler/visitors/generateOperationId.ts#L5)). We could choose to do something similar.
```

Comment:
Using just the `id` also wont reveal the true culprit since there can be variables assocaited with the query. 
```
{
  "id": "some_hash_generated_for_the_query",
  "variables": {
    "id": "foo"
  }
}
```

```
{
  "id": "some_hash_generated_for_the_query",
  "variables": {
    "id": "bar",
     name: "baz"
  }
}

```

Both are different and would result in different query cost and server response time.

Reply:
Right. How do you feel about _only_ handling queries with _operation name_ and treating the rest as "dumb" or at least "anonymous"?

Reply:
Yeah thats a sane approach, We can also log a warning from the agent stating having operationName would help in identifying issues faster.